components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/action/PreIssueAccessTokenRequestBuilder.java

```diff
@@ -0,0 +1,352 @@
+
+    public static final String CLAIMS_PATH_PREFIX = "/accessToken/claims/";
+    public static final String SCOPES_PATH_PREFIX = "/accessToken/scopes/";
+    private static final Log LOG = LogFactory.getLog(PreIssueAccessTokenRequestBuilder.class);
+
+    @Override
+    public ActionType getSupportedActionType() {
+
+        return ActionType.PRE_ISSUE_ACCESS_TOKEN;
+    }
+
+    @Override
+    public ActionExecutionRequest buildActionExecutionRequest(Map<String, Object> eventContext)
+            throws ActionExecutionRequestBuilderException {
+
+        OAuthTokenReqMessageContext tokenMessageContext =
+                (OAuthTokenReqMessageContext) eventContext.get("tokenMessageContext");
+
+        Map<String, Object> additionalClaimsToAddToToken = getAdditionalClaimsToAddToToken(tokenMessageContext);
+
+        ActionExecutionRequest.Builder actionRequestBuilder = new ActionExecutionRequest.Builder();
+        actionRequestBuilder.actionType(getSupportedActionType());
+        actionRequestBuilder.event(getEvent(tokenMessageContext, additionalClaimsToAddToToken));
+        actionRequestBuilder.allowedOperations(getAllowedOperations(additionalClaimsToAddToToken));
+
+        return actionRequestBuilder.build();
+    }
+
+    private Event getEvent(OAuthTokenReqMessageContext tokenMessageContext, Map<String, Object> claimsToAdd)
+            throws ActionExecutionRequestBuilderException {
+
+        OAuth2AccessTokenReqDTO tokenReqDTO = tokenMessageContext.getOauth2AccessTokenReqDTO();
+        AuthenticatedUser authorizedUser = tokenMessageContext.getAuthorizedUser();
+
+        PreIssueAccessTokenEvent.Builder eventBuilder = new PreIssueAccessTokenEvent.Builder();
+        eventBuilder.tenant(new Tenant(String.valueOf(IdentityTenantUtil.getTenantId(tokenReqDTO.getTenantDomain())),
+                tokenReqDTO.getTenantDomain()));
+
+        boolean isAuthorizedForUser = isAccessTokenAuthorizedForUser(tokenReqDTO.getGrantType(), tokenMessageContext);
+        if (isAuthorizedForUser) {
+            setUserForEventBuilder(eventBuilder, authorizedUser, tokenReqDTO.getClientId(), tokenReqDTO.getGrantType());
+            setOrganizationForEventBuilder(eventBuilder, authorizedUser, tokenReqDTO.getClientId(),
+                    tokenReqDTO.getGrantType());
+            eventBuilder.userStore(new UserStore(authorizedUser.getUserStoreDomain()));
+        }
+
+        eventBuilder.accessToken(getAccessToken(tokenMessageContext, claimsToAdd));
+        eventBuilder.request(getRequest(tokenReqDTO));
+
+        return eventBuilder.build();
+    }
+
+    private void setUserForEventBuilder(PreIssueAccessTokenEvent.Builder eventBuilder, AuthenticatedUser user,
+                                        String clientID, String grantType) {
+
+        try {
+            eventBuilder.user(new User(user.getUserId()));
+        } catch (UserIdNotFoundException e) {
+            if (LOG.isDebugEnabled()) {
```

Comment:
Are we not breaking the flow even if the 'user id' is not found when the access token us authorized for user?

Reply:
No, IMO user id is not found it should fall back to a username or something. Need to verify this behaviour more and improve.

Reply:
Added a todo so we can improve later